src/main/kotlin/org/rust/lang/core/CFGBuilder.kt

```diff
@@ -0,0 +1,348 @@
+        //        |                  |
+        //        v 1                v 1
+        //      [cond]             [cond]
+        //        |                  |
+        //       / \                / \
+        //      /   \              /   \
+        //     v 2   v 3          v 2   *
+        //   [then][else]       [then]  |
+        //     |     |            |     |
+        //     v 4   v 5          v 3   v 4
+        //     [ifExpr]          [ifExpr]
+        //
+        val conditionExit = process(ifExpr.condition?.expr, pred)
+        val thenExit = process(ifExpr.block, conditionExit)
+        val elseBranch = ifExpr.elseBranch
+
+        if (elseBranch != null) {
+            val elseExit = process(elseBranch.block, conditionExit)
+            finishWith { addAstNode(ifExpr, listOf(thenExit, elseExit)) }
+        } else {
+            finishWith { addAstNode(ifExpr, listOf(conditionExit, thenExit)) }
+        }
+    }
+
+    override fun visitWhileExpr(whileExpr: RsWhileExpr) {
+        //
+        //         [pred]
+        //           |
+        //           v 1
+        //       [loopback] <--+ 5
+        //           |         |
+        //           v 2       |
+        //   +-----[cond]      |
+        //   |       |         |
+        //   |       v 4       |
+        //   |     [body] -----+
+        //   v 3
+        // [whileExpr]
+        //
+        val loopBack = addDummyNode(listOf(pred))
+        val exprExit = addAstNode(whileExpr, emptyList())
+        val loopScope = LoopScope(whileExpr, loopBack, exprExit)
+
+        withLoopScope(loopScope) {
+            val conditionExit = process(whileExpr.condition?.expr, loopBack)
+            addContainedEdge(conditionExit, exprExit)
+
+            val bodyExit = process(whileExpr.block, conditionExit)
+            addContainedEdge(bodyExit, loopBack)
+        }
+
+        finishWith(exprExit)
+    }
+
+    override fun visitLoopExpr(loopExpr: RsLoopExpr) {
+        //
+        //     [pred]
+        //       |
+        //       v 1
+        //   [loopback] <---+
+        //       |      4   |
+        //       v 3        |
+        //     [body] ------+
+        //
+        //   [loopExpr] 2
+        //
+        val loopBack = addDummyNode(listOf(pred))
+        val exprExit = addAstNode(loopExpr, emptyList())
+        val loopScope = LoopScope(loopExpr, loopBack, exprExit)
+
+        withLoopScope(loopScope) {
+            val bodyExit = process(loopExpr.block, loopBack)
+            addContainedEdge(bodyExit, loopBack)
+        }
+
+        finishWith(exprExit)
+    }
+
+    override fun visitForExpr(forExpr: RsForExpr) {
+        val loopBack = addDummyNode(listOf(pred))
+        val exprExit = addAstNode(forExpr, emptyList())
+        val loopScope = LoopScope(forExpr, loopBack, exprExit)
+
+        withLoopScope(loopScope) {
+            val bodyExit = process(forExpr.block, loopBack)
+            addContainedEdge(bodyExit, loopBack)
```

Comment:
Looks like we processing for expression just like it is loop expression, that is incorrect.